clippy_lints/src/index_refutable_slice.rs

```diff
@@ -0,0 +1,278 @@
+                // The values need to use the `ref` keyword if they can't be coppied.
+                // This will need to be adjusted if the lint want to support multable access in the future
+                let needs_ref = !bound_ty.is_ref()
+                    || !cx
+                        .tcx
+                        .lang_items()
+                        .copy_trait()
+                        .map_or(false, |trait_id| implements_trait(cx, inner_ty, trait_id, &[]));
+
+                let slice_info = slices
+                    .entry(value_hir_id)
+                    .or_insert_with(|| SliceLintInformation::new(ident, needs_ref));
+                slice_info.pattern_spans.push(pat.span);
+            }
+        }
+    });
+
+    (!slices.is_empty()).then(|| slices)
+}
+
+fn lint_slice(cx: &LateContext<'_>, slice: &SliceLintInformation) {
+    let used_indices = slice
+        .index_use
+        .iter()
+        .map(|(index, _)| *index)
+        .collect::<FxHashSet<_>>();
+
+    let value_name = |index| format!("{}_{}", slice.ident.name, index);
+
+    if let Some(max_index) = used_indices.iter().max() {
+        let opt_ref = if slice.needs_ref { "ref " } else { "" };
+        let pat_sugg_idents = (0..=*max_index)
+            .map(|index| {
+                if used_indices.contains(&index) {
+                    format!("{}{}", opt_ref, value_name(index))
+                } else {
+                    "_".to_string()
+                }
+            })
+            .collect::<Vec<_>>();
+        let pat_sugg = format!("[{}, ..]", pat_sugg_idents.join(", "));
+
+        span_lint_and_then(
+            cx,
+            INDEX_REFUTABLE_SLICE,
+            slice.ident.span,
+            "this binding can be a slice pattern to avoid indexing",
+            |diag| {
+                diag.multipart_suggestion(
+                    "try using a slice pattern here",
+                    slice
+                        .pattern_spans
+                        .iter()
+                        .map(|span| (*span, pat_sugg.clone()))
+                        .collect(),
+                    Applicability::MaybeIncorrect,
+                );
+
+                diag.multipart_suggestion(
+                    "and replacing the index expressions here",
+                    slice
+                        .index_use
+                        .iter()
+                        .map(|(index, span)| (*span, value_name(*index)))
+                        .collect(),
+                    Applicability::MaybeIncorrect,
+                );
+
+                // I thought about adding a note to the lint message to inform the user that these
+                // refactorings will remove the index expression. However, I decided against this,
+                // as `filter_lintable_slices` will only return slices where all access indices are
+                // known at compile time. The removal should therefore not have any side effects.
```

Comment:
Yes, this is referring to the expression inside the index. I was referring to a case like this:
```rs
fn give_index() -> usize {
    println!("I have side effects");
    println!("I print stuff");
    println!("I important function");
    1
}

fn main() {
    // ...
    if let Some(ref slice) = slice {
        println!("{:?}", slice[give_index()]);
    //                         ^^^^^^^^^^^^
    // Removing this also removes the side effects of `give_index`. This is
    // not directly stated by "and replacing the index expressions here".
    // But it's very much implied and wouln't be linted anyway as `give_index`
    // is not a const expression at compiletime
    // ...
}
```

Reply:
I see. So it sounds like "I thought about doing X in order to mitigate problem Y, but then I realized problem Y doesn't exist because Z". Perhaps reduce that to "Problem Y doesn't exist because Z". But I don't mean to linger on this.